compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -851,40 +902,387 @@ installDepsAndInitPackageDb opts (InitPkgDb shouldInit) =
+  the multi-package.yaml is not required to be in the same directory as any given package. We only require it is at or above
+  (in the directory tree) the package you are building. multi-build will first look for this file before attempting to build.
+
+  The process is as follows:
+  - Search for the multi-package.yaml file by traversing up the directory tree to root until we are at a directory containing the file.
+  - Set up own shake environment, defer to it immediately with the project path
+  - Within the shake rule:
+  - Partially read the daml.yaml file, looking for data dependencies, and some other details needed for caching
+  - Intercept the packages data dependencies with the packages we have listed in multi-package.yaml.
+  - Call to shake to build all the above subset of data-dependencies, each will return data about the DAR that exists (or was built)
+  - If we already have a DAR, check its staleness. Staleness is defined by any source files being different
+      or any of the package ids of souce dependencies being different between what is currently built, and what is embedded in our dar.
+  - If either are different, call to the SDK via CLI to build our own DAR.
+  - Get the package ID of the DAR we have created, and return it to shake, which can use it as a caching key.
+
+  Due to shakes caching behaviour, this will automatically discover the build order for free, by doing a depth first search over the tree then
+    caching any "nodes" we've already seen entirely within Shake - our rule will not be called twice for any package within the tree for any given multi-build.
+  
+  We currently rely on shake to find cycles, how its errors includes too much information about internals, so we'll need to implement our own cycle detection.
+-}
+multiPackageBuildEffect
+  :: (FilePath -> IO FilePath)
+  -> Maybe PackageConfigFields -- Nothing signifies build all
+  -> MultiPackageConfigFields
+  -> ProjectOpts
+  -> Options
+  -> Maybe FilePath
+  -> IncrementalBuild
+  -> InitPkgDb
+  -> MultiPackageNoCache
+  -> IO ()
+multiPackageBuildEffect relativize mPkgConfig multiPackageConfig projectOpts opts mbOutFile incrementalBuild initPkgDb noCache = do
+  vfs <- makeVFSHandle
+  loggerH <- getLogger opts "multi-package build"
+  cDir <- getCurrentDirectory
+  assistantPath <- getEnv "DAML_ASSISTANT"
+  -- Must drop DAML_PROJECT from env var so it can be repopulated based on `cwd`
+  assistantEnv <- filter ((/="DAML_PROJECT") . fst) <$> getEnvironment
+
+  buildableDataDepsMapping <- fmap Map.fromList $ for (mpPackagePaths multiPackageConfig) $ \path -> do
+    darPath <- darPathFromDamlYaml path
+    pure (darPath, path)
+
+  let assistantRunner = AssistantRunner $ \location args ->
+        withCreateProcess ((proc assistantPath args) {cwd = Just location, env = Just assistantEnv})
+          $ \_ _ _ p -> void $ waitForProcess p
```

Comment:
Can we handle the assistant running and exiting badly, and maybe print stderr/stdout when that happens with an error message? If something goes wrong internally, it'd be good for us when the user has an error message to give us so we can diagnose things.

Reply:
The process inherits the handles from the main process, so its stdout and stderr do get printed, but the main process doesn't halt, so i'll fix that.